### ISSUE TITLE
ML-KEM doesn't perform encapsulation key check

### ISSUE THREAD

**Issue by ueno** [open-quantum-safe/liboqs#1951]
**Describe the bug**
When encapsulating, FIPS 204 7.1 [suggests](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.203.pdf#page=36) checking input key so any integers in the encoded public key is in the range [0, q - 1]. However, the current liboqs code seems to omit this check.

**To Reproduce**
1. Compile the attached [test-kem.c](https://github.com/user-attachments/files/17372939/test-kem.c.txt) with:  `gcc -o test-kem test-kem.c $(pkg-config liboqs --cflags --libs)`
2. Run `test-kem`
3. You will see: `test-kem: test-kem.c:48: main: Assertion 'rc != OQS_SUCCESS' failed.`

**Expected behavior**
The program should exit normally.

**Environment (please complete the following information):**
 - OS: Fedora Linux 40
 - OpenSSL version 3.2.2
 - Compiler version used gcc 14.2.1
 - Build variables used `-DBUILD_SHARED_LIBS=ON -DOQS_USE_AES_OPENSSL=ON -DOQS_USE_AES_INSTRUCTIONS=OFF -DOQS_DIST_BUILD=ON -DOQS_ALGS_ENABLED=STD_IANA -DOQS_USE_SHA3_OPENSSL=ON -DOQS_DLOPEN_OPENSSL=ON -DCMAKE_BUILD_TYPE=Debug -LAH`
 - liboqs version 0.11.0

**Additional context**
This was found when running a tlsfuzzer test [script](https://github.com/tlsfuzzer/tlsfuzzer/blob/ml-kem/scripts/test-tls13-mlkem.py#L258) currently in development, against gnutls-serv (in [MR](https://gitlab.com/gnutls/gnutls/-/merge_requests/1883)).

**Comment by SWilson4** [comment]
Thanks for the report, @ueno. We pull our ML-KEM implementation from the [pq-crystals/kyber](https://github.com/pq-crystals/kyber) repo. It seems to me that it would be better to add the optional checks upstream and then integrate the updated code into liboqs. Would you agree with that assessment?

As an aside, I don't believe we support the `STD_IANA` value for the `OQS_ALGS_ENABLED` variable—I assume this is something from a fork?

**Comment by SWilson4** [comment]
Also tagging @bhess as he is the one most familiar with the CRYSTALS algorithms.

**Comment by nidhidamodaran** [comment]
Can you please help me understand whether this patch to liboqs will be able to identify bad keys

`+int poly_frombytes(poly *r, const uint8_t a[KYBER_POLYBYTES])
 {
   unsigned int i;
   for(i=0;i<KYBER_N/2;i++) {
     r->coeffs[2*i]   = ((a[3*i+0] >> 0) | ((uint16_t)a[3*i+1] << 8)) & 0xFFF;
     r->coeffs[2*i+1] = ((a[3*i+1] >> 4) | ((uint16_t)a[3*i+2] << 4)) & 0xFFF;
+    if((r->coeffs[2*i] >= KYBER_Q) || (r->coeffs[2*i+1]  >= KYBER_Q))
+       return -1;
   }
+  return 0;
 }`

**Comment by tomato42** [comment]
@nidhidamodaran on first look, yes, it does look like the thing that's missing

**Comment by SWilson4** [comment]
> Can you please help me understand whether this patch to liboqs will be able to identify bad keys
> 
> ```
> +int poly_frombytes(poly *r, const uint8_t a[KYBER_POLYBYTES])
>  {
>    unsigned int i;
>    for(i=0;i<KYBER_N/2;i++) {
>      r->coeffs[2*i]   = ((a[3*i+0] >> 0) | ((uint16_t)a[3*i+1] << 8)) & 0xFFF;
>      r->coeffs[2*i+1] = ((a[3*i+1] >> 4) | ((uint16_t)a[3*i+2] << 4)) & 0xFFF;
> +    if((r->coeffs[2*i] >= KYBER_Q) || (r->coeffs[2*i+1]  >= KYBER_Q))
> +       return -1;
>    }
> +  return 0;
>  }
> ```

This patch is not a good idea for a couple of reasons.

1. **Most importantly**, the `>=` operations is not constant time and may leak information about whatever data it operates on. The `poly_frombytes` function is used to process the secret key, so this is a problem.
2. This patch only works for the `ref`, non-optimized code. There's a good chance you're not running this code at all.

**Comment by tomato42** [comment]
> Most importantly, the >= operation is not constant time and may leak information about whatever data it operates on. The poly_frombytes function is used to process the secret key, so this is a problem.

when would it be operating on a secret values though?

either it's used for checking input for decapsulation: that's public
or it's used for sanity checking output of encapsulation: that will be public...

**Comment by SWilson4** [comment]
> > Most importantly, the >= operation is not constant time and may leak information about whatever data it operates on. The poly_frombytes function is used to process the secret key, so this is a problem.
> 
> when would it be operating on a secret values though?
> 
> either it's used for checking input for decapsulation: that's public or it's used for sanity checking output of encapsulation: that will be public...

The `poly_frombytes` function is used throughout the Kyber / ML-KEM reference code. In particular, it's used (via the calls `pack_sk` <- `polyvec_frombytes` <- `poly_frombytes` here to process the secret key at generation time.

**Comment by nidhidamodaran** [comment]
Right, for my use case i have defined a different API available only in ref implementation,

**Comment by SWilson4** [comment]
> The `poly_frombytes` function is used throughout the Kyber / ML-KEM reference code. In particular, it's used (via the calls `pack_sk` <- `polyvec_frombytes` <- `poly_frombytes`) [here](https://github.com/open-quantum-safe/liboqs/blob/416778ecc2cd246d02c5684486a7a2489a9d8fbf/src/kem/ml_kem/pqcrystals-kyber-standard_ml-kem-768_ref/indcpa.c#L241) to process the secret key at generation time.

Sorry, this trace isn't what I meant to send. Give me a second...
